### PR TITLE
Fix USING clause for JOINs

### LINKED ISSUE #2311: Fix USING clause for JOINs
Columns that appear in a `USING` clause should not require qualification. This is particularly useful for outer joins, because the non-qualified reference always references the non-null value. It is not clear if qualifying reference is allowed: PostgreSQL allows it but Oracle does not. Allowing it is useful in order to do `IS NULL` checks for outer joins, especially when no other columns are available.

```
presto> select x from (values (1)) a (x) join (values (1)) b (x) using (x);
Query 20150202_160415_12625_wu6ji failed: Column 'x' is ambiguous
```

Additionally, because `USING` makes them the same column, the column should only appear once in the output descriptor:

```
presto> select * from (values (1)) a (x) join (values (1)) b (x) using (x);
 x | x 
---+---
 1 | 1 
(1 row)
```

Correct behavior for the above:

```
postgresql=# select x from (values (1)) a (x) join (values (1)) b (x) using (x);
 x 

---
 1
(1 row)

postgresql=# select * from (values (1)) a (x) join (values (1)) b (x) using (x);
 x 

---
 1
(1 row)
```

Additional test cases that should return the value from the outside of the join:

``` sql
SELECT x FROM (VALUES (1), (2)) a (x) LEFT JOIN (VALUES (1)) b (x) USING (x);
SELECT x FROM (VALUES (1)) a (x) RIGHT JOIN (VALUES (1), (2)) b (x) USING (x);
```

### PR DESCRIPTION
**Pull request by kokosing**
Fix USING clause for JOINs

Fixes #2311

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 1]
```diff
@@ -806,6 +806,13 @@ protected Scope visitJoin(Join node, Scope scope)
 
                 addCoercionForJoinCriteria(node, leftExpression, rightExpression);
                 expressions.add(new ComparisonExpression(EQUAL, leftExpression, rightExpression));
+
+                if (node.getType() == Join.Type.LEFT) {
+                    output = createScope(node, scope, left.getRelationType().joinWith(right.hide(rightExpression).getRelationType()));
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: I don't think this is the right semantic. The fields are not supposed to be "hidden". They should both be replaced by a single column with no relation alias.

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 2]
```diff
@@ -2004,10 +2004,44 @@ public void testJoinWithInferredFalseJoinClause()
...
+                "VALUES (1, 0, 3), (4, 3, 0)");
+
+        assertQuery(
+                "SELECT * FROM (VALUES (0, 1, 2), (3, 6, 5)) a(x1, y, z1) LEFT OUTER JOIN (VALUES (3, 1, 5), (0, 4, 2)) b(x2, y, z2) USING (y)",
+                "VALUES (0, 1, 2, 3, 5), (3, 6, 5, null, null)");
+        assertQuery(
+                "SELECT y, x1, x2 FROM (VALUES (0, 1, 2), (3, 6, 5)) a(x1, y, z1) LEFT OUTER JOIN (VALUES (3, 1, 5), (0, 4, 2)) b(x2, y, z2) USING (y)",
+                "VALUES (1, 0, 3), (6, 3, null)");
+
+        assertQuery(
```
Comment (electrum) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: Also test `FULL OUTER JOIN`:
```sql
SELECT * FROM (VALUES (0, 1, 2), (3, 6, 5)) a(x1, y, z1) FULL OUTER JOIN (VALUES (3, 1, 5), (0, 4, 2)) b(x2, y, z2) USING (y);
VALUES (0, 2, 3, 1, 5), (6, 3, 5, null, null), (null, null, 0, 4, 2);

SELECT y, x1, x2 FROM (VALUES (0, 1, 2), (3, null, 5)) a(x1, y, z1) FULL OUTER JOIN (VALUES (3, 1, 5), (0, 4, 2)) b(x2, y, z2) USING (y);
VALUES (1, 0, 3), (null, 3, null), (4, null, 0);

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 1 above)
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Although I'm not sure if this is correct:
```
dphillips=# SELECT a.*, b.* FROM (VALUES (1), (2)) a (x) FULL JOIN (VALUES (1), (3)) b (x) USING (x);
 x | x 
---+---
 1 | 1
 2 |  
   | 3
(3 rows)
```
I think they should not be resolved as part of `a.*`, although it is useful if they are available as a hidden column.

Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: I'd prefer to stick to standard sql semantics to start with. We can evaluate whether we want or need that later.

Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: I typically use postgres to check the how given query behaves as I assume it has the best support of ANSI SQL within open source databases (am I right?). And it looks like it more complicated than I thought, please the below:

```
postgres=# select column1, t1.column1, t2.column1, * from t1 full outer join t2 using(column1);
 column1 | column1 | column1 | column1 | column2 | column2 
---------+---------+---------+---------+---------+---------
       1 |       1 |       1 |       1 |      -1 |     -11
       2 |       2 |         |       2 |      -2 |        
         |         |         |         |      -4 |        
       3 |         |       3 |       3 |         |     -33
         |         |         |         |         |     -44
(5 rows)

```

It looks like that `a.x` and `b.x` are hidden and `a` is equal to an expression `case a.x IS NULL THEN b.x ELSE a.x END`.

Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: I need to re-read that part of the spec, but if I recall correctly, then qualified columns become unavailable in the scope and the join columns (there could be multiple), are equivalent to coalesce(l.x, r.x)

Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: PostgreSQL has decent support for ANSI sql, but it deviates in a few places.

**Comment by electrum** [comment]
Good point about multiple joins. Add a test that does two FULL joins where the non-null join column value comes from all three tables.

**Comment by wenleix** [comment]
@kokosing Thank you for the great work!!
Today when we do `A FULL OUTER JOIN B FULL OUTER JOIN C` on the same column, Presto will reshuffle the results of `A FULL OUTER JOIN B` before joining it with C. This is (I believe) because the partitioning information will miss after the `COALESCE`.

With the full support of `USING` Clause, will Presto avoid reshuffling the data after each `OUTER JOIN?` (like what it did in `LEFT OUTER JOIN`)  ?

**Comment by kokosing** [comment]
@wenleix Do you mean shuffling rows or columns? This PR is about setting proper order of columns where  USING is used, and order is defined by the standard. AFAIR, first columns from USING then left columns and the right columns.

This PR is out-dated and needs more effort to make to make it work correctly. Hope to get back to it in couple of weeks.

**Comment by wenleix** [comment]
@kokosing Shuffling rows. Sorry I should use word `Exchange` since this is the term in Presto :).

For example, for the following `LEFTR OUTER JOIN` query, Presto knows that the data is already partitioned by join key, thus there is not remote exchange for following joins:

```

WITH t1 AS (
SELECT orderkey, linenumber FROM lineitem WHERE linenumber = 1), 
t2 AS (
SELECT orderkey, quantity FROM lineitem WHERE linenumber = 2), 
t3 AS (
SELECT orderkey, extendedprice FROM lineitem WHERE linenumber = 3), 
t4 AS (
SELECT orderkey, discount FROM lineitem WHERE linenumber = 4), 
t5 AS (
SELECT orderkey, tax FROM lineitem WHERE linenumber = 5)

SELECT * 
FROM t1 LEFT OUTER JOIN t2 ON t1.orderkey = t2.orderkey
  LEFT OUTER JOIN t3 ON t1.orderkey = t3.orderkey
  LEFT OUTER JOIN t4 ON t1.orderkey = t4.orderkey
  LEFT OUTER JOIN t5 ON t1.orderkey = t5.orderkey
```

Here is the query plan:
screen shot 2017-02-14 at 10 26 16


However, this doesn't hold for the following `FULL OUTER JOIN` query: 
```
WITH t1 AS (
SELECT orderkey, linenumber FROM lineitem WHERE linenumber = 1), 
t2 AS (
SELECT orderkey, quantity FROM lineitem WHERE linenumber = 2), 
t3 AS (
SELECT orderkey, extendedprice FROM lineitem WHERE linenumber = 3), 
t4 AS (
SELECT orderkey, discount FROM lineitem WHERE linenumber = 4), 
t5 AS (
SELECT orderkey, tax FROM lineitem WHERE linenumber = 5),

t12 AS (
  SELECT coalesce(t1.orderkey, t2.orderkey) as orderkey
  FROM t1 FULL OUTER JOIN t2 ON t1.orderkey = t2.orderkey
),
t123 AS (
  SELECT coalesce(t12.orderkey, t3.orderkey) as orderkey
  FROM t12 FULL OUTER JOIN t3 ON t12.orderkey = t3.orderkey
),
t1234 AS (
  SELECT coalesce(t123.orderkey, t4.orderkey) as orderkey
  FROM t123 FULL OUTER JOIN t4 ON t123.orderkey = t4.orderkey
),
t12345 AS (
  SELECT coalesce(t1234.orderkey, t5.orderkey) as orderkey
  FROM t1234 FULL OUTER JOIN t5 ON t1234.orderkey = t5.orderkey
)

SELECT * FROM t12345

```

As you can note from the Presto query plan, it will do remote exchange after each join:

screen shot 2017-02-13 at 22 44 16


It's quite difficult for query planner to infer `COALESCE(t1.orderkey, t2.orderkey)` is actually partitioned. However, with the `USING` clause, it should be able to infer the join key is partitioned and thus avoids the further remote exchange operation.

**Comment by kokosing** [comment]
Nice. That would mean that this issue would have higher priority then.

Although, I don't like the idea that depending on whether user use `USING(col)` or `(ON x.col = y.col)` he would completely different plan, while there should be almost the same. Is it possible to push down `COALESCE(t1.orderkey, t2.orderkey)` somehow and then partition on it?

**Comment by kokosing** [comment]
CC: @pnowojski

**Comment by pnowojski** [comment]
@wenleix your first plan is `LEFT OUTER JOIN` performed by broadcast join (side node, when right table is comparable in size to left table, it is better to use default distributed join). Regarding your second query it is impossible to perform broadcast join for FULL OUTER or RIGHT OUTER . That's why presto will always choose distributed join in such case, and that's the reason why there are exchange nodes between joins.

**Comment by wenleix** [comment]
@pnowojski : I run the query with `distrubited_join` properties set to true. So both plan should be using distributed join :).

The `LEFT OUTER JOIN` avoids do `Exchange` on the left side because it knows after the JOIN, the output is already partitioned on the JOIN key. But this information is missing for `FULL OUTER JOIN` because of the `COALESCE(t1.orderkey, t2.orderkey)` operation.

@kokosing : Yeah, ideally the engine should be able to infer `COALESCE(t1.orderkey, t2.orderkey)` is partitioned with an `FULL OUTER JOIN`, however after discussing with @haozhun , this seems to be quite difficult for the engine to reason right now. And `USING` clause is much easier for the engine to understand.

**Comment by kokosing** [comment]
ok, I will try to prioritize this work and fix this issue in the near future so you could continue you work related to join optimization.

**Comment by martint** [comment]
USING is just a syntactic feature. I don't see how it would help, since it would end up being desugared into that COALESCE call. I don't think we should be introducing a new type of join.

**Comment by pnowojski** [comment]
@wenleix ok, I get it now

**Comment by kokosing** [comment]
I am closing this as I do not actively work on this.

**Pull request closed without merging**